src/Workspaces/Remote/Core/AbstractAssetProvider.cs

```diff
@@ -74,26 +74,28 @@ public async Task<SolutionInfo> CreateSolutionInfoAsync(Checksum solutionChecksu
             analyzerConfigDocumentInfos,
             hostObjectType: null); // TODO: https://github.com/dotnet/roslyn/issues/62804
 
-        async Task<ImmutableArray<DocumentInfo>> CreateDocumentInfosAsync(ChecksumCollection documentChecksums)
+        async Task<ImmutableArray<DocumentInfo>> CreateDocumentInfosAsync(ChecksumsAndIds<DocumentId> checksumsAndIds)
         {
-            using var _ = ArrayBuilder<DocumentInfo>.GetInstance(documentChecksums.Count, out var documentInfos);
+            using var _ = ArrayBuilder<DocumentInfo>.GetInstance(checksumsAndIds.Length, out var documentInfos);
 
-            foreach (var documentChecksum in documentChecksums)
+            foreach (var (documentChecksum, documentId) in checksumsAndIds)
             {
                 cancellationToken.ThrowIfCancellationRequested();
-                documentInfos.Add(await CreateDocumentInfoAsync(projectId, documentChecksum, cancellationToken).ConfigureAwait(false));
+                documentInfos.Add(await CreateDocumentInfoAsync(documentId, documentChecksum, cancellationToken).ConfigureAwait(false));
             }
 
             return documentInfos.ToImmutableAndClear();
         }
     }
 
     public async Task<DocumentInfo> CreateDocumentInfoAsync(
-        ProjectId projectId, Checksum documentChecksum, CancellationToken cancellationToken)
+        DocumentId documentId, Checksum documentChecksum, CancellationToken cancellationToken)
     {
-        var documentSnapshot = await GetAssetAsync<DocumentStateChecksums>(hintProject: projectId, documentChecksum, cancellationToken).ConfigureAwait(false);
-        var attributes = await GetAssetAsync<DocumentInfo.DocumentAttributes>(hintProject: projectId, documentSnapshot.Info, cancellationToken).ConfigureAwait(false);
-        var serializableSourceText = await GetAssetAsync<SerializableSourceText>(hintProject: projectId, documentSnapshot.Text, cancellationToken).ConfigureAwait(false);
+        var documentSnapshot = await GetAssetAsync<DocumentStateChecksums>(assetHint: documentId, documentChecksum, cancellationToken).ConfigureAwait(false);
```

Comment:
better than before.  we can get the document checksums now without having to search a whoel project for it.